### DISCUSSION TITLE
One trpc (nextjs) backend w/ multiple frontends

### DISCUSSION THREAD

**Discussion by spiritanand** [trpc/trpc#5943] [Q&A] [answered]
Is it possible to have a single nextjs backend powered by the api routes with trpc and use that endpoints (not the function) to actually be called of multiple different frontends?

Please let me know if there is way for that

**Comment by BeBoRE** [comment]
You can use a mono repo for this

**Comment by spiritanand** [comment]
Thanks @BeBoRE, but is it possible to have my nextjs backend hosted (powered by tRPC) and then use those enpoints as API's. I am guessing it should be possible, but definitely not recommended, since we would be losing on the typesafety right. 
I was just wondering if there is a way around it?

**Reply by Floffah** [reply, marked as answer]
the structure you want might be similar to what i do in [this repo](https://github.com/Floffah/dndnotes). see apps/api and apps/web

**Reply by spiritanand** [reply]
@Floffah this is what I was talking about, though I had a few questions:

1. a thing that strikes me is that we will need to init a trpc route on every project we create, why is that (`apps/web/src/app/api/[api]/route.ts`)

2. Also why do we need to create new trpc (`apps/web/src/lib/api/index.ts`)?

3. Intereseting that you did not make the api (trpc) a package rather it is in the apps. Any reason for that?

**Reply by Floffah** [reply]
> @Floffah this is what I was talking about, though I had a few questions:
> 
> 1. a thing that strikes me is that we will need to init a trpc route on every project we create, why is that (`apps/web/src/app/api/[api]/route.ts`)
> 2. Also why do we need to create new trpc (`apps/web/src/lib/api/index.ts`)?
> 3. Intereseting that you did not make the api (trpc) a package rather it is in the apps. Any reason for that?

1. You don't necessarily need to have a trpc endpoint on every next project, but if you want to keep things on the same domain, you can - it will always be the same API. The way I do it is by having the endpoint on whatever the main package is (e.g. the one deployed to the root domain) and pointing the trpc client on each next project to it through environment variables. keeps costs low. I have an example of this on the v1 branch of that repo: apps/discord-activity calls the trpc endpoint on apps/web. Your CORS situation may be different though.
2. It's probably able to be argued that it is a good practice to create a new trpc client on each project, but I don't think it matters all too much - it's just what you feel has the best DX. The only way to share this is to create a separate package just for the client, as we can't import the API package on the client outwith types. I ended up doing this because both my frontends authenticated separately: one through cookies and one through an injected header so I made two different clients with their own behaviour.
3. My perfectionism 😄 it's just in case I ever want to deploy the trpc package to its own server. Although, in my mind they are both 'apps' as they are the only two user-facing packages

**Reply by spiritanand** [reply]
Thank you so much @Floffah 🫡

**Comment by Rezanmonis** [comment]
We are looking for frontend and backend developers as our teammates for our edtech project rezanmonis@gmail.com
